cni/pkg/ebpf/app/ambient_redirect.bpf.c

```diff
@@ -0,0 +1,262 @@
+{
+    return bpf_map_lookup_elem(&app_info, &ipv4);
+}
+
+static __inline struct ztunnel_info * get_ztunnel_info()
+{
+    uint32_t key = 0;
+    return bpf_map_lookup_elem(&ztunnel_info, &key);
+}
+
+// For app pod veth pair(in host ns) ingress hook
+SEC("tc")
+int app_outbound(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct ethhdr *eth = data;
+    struct iphdr *iph;
+    __u32 outbound_cb = OUTBOUND_CB;
+    struct ztunnel_info *zi = get_ztunnel_info();
+
+    if (!zi || zi->ifindex == 0)
+        return TC_ACT_OK;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP)
+        return TC_ACT_OK;
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = outbound_cb;
+
+    dbg("outbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For app pod veth pair(in host ns) egress hook
+SEC("tc")
+int app_inbound(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct ethhdr  *eth = data;
+    struct iphdr *iph = NULL;
+    struct ztunnel_info *zi = NULL;
+    __u32 inbound_cb = INBOUND_CB;
+
+    if (skb->mark == ZTUNNEL_BYPASS_MARK)
+        return TC_ACT_OK;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP)
+        return TC_ACT_OK;
+
+
+    zi = get_ztunnel_info();
+
+    if (!zi || zi->ifindex == 0) {
+        dbg("inbound_to_ztunnel unable to retrieve ztunnel_info\n");
+        return TC_ACT_OK;
+    }
+
+    __builtin_memcpy(eth->h_dest, zi->mac_addr, ETH_ALEN);
+
+    skb->cb[4] = inbound_cb;
+
+    dbg("inbound redirect to ztunnel ifidx: %u\n", zi->ifindex);
+    return bpf_redirect(zi->ifindex, 0);
+}
+
+// For ztunnel pod veth pair(in host ns) ingress hook
+SEC("tc")
+int ztunnel_host_ingress(struct __sk_buff *skb)
+{
+    void *data = (void *)(long)skb->data;
+    struct ethhdr  *eth = data;
+    void *data_end = (void *)(long)skb->data_end;
+    struct iphdr *iph = NULL;
+    struct app_info *pi = NULL;
+
+    if (data + sizeof(*eth) > data_end)
+        return TC_ACT_OK;
+
+    // TODO: support ipv6
+    if (eth->h_proto != bpf_htons(ETH_P_IP))
+        return TC_ACT_OK;
+
+    iph = data + sizeof(*eth);
+    if (data + sizeof(*eth) + sizeof(*iph) > data_end)
+        return TC_ACT_OK;
+
+    if (iph->protocol != IPPROTO_TCP )
```

Comment:
One question: Any particular reason to enforce this?   I.e. if the packet isn't IP, why not just forward it?  We definitely want to support UDP in future.

Reply:
we can release this enforcemernt.